### DISCUSSION TITLE
executeJob() 'outside of map area' return differents for DatabaseRenderer and DirectRenderer

### DISCUSSION THREAD

**Discussion by nono303** [mapsforge/mapsforge#1300] [Q&A] [answered]
Hi,
Using mapsforge on [telemaxx/mapsforgesrv](https://github.com/telemaxx/mapsforgesrv), we currently face a strange behavior on `Renderer.executeJob` results for _outside of map area_ cases _(see [#29](https://github.com/telemaxx/mapsforgesrv/issues/29))_
 - with `DatabaseRenderer` a **blank TileBitmap** is returned
https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/layer/renderer/DatabaseRenderer.java#L131
```
            // outside of map area with background defined:
            return createBackgroundBitmap(renderContext);
```
 - with `DirectRenderer` **null**  is returned
https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/layer/renderer/DirectRenderer.java#L119
```
            // outside of map area
            return null;
```
Is these different behaviors (on simply a different implementation) are normal or it's an issue?

**Comment by devemux86** [comment]
`DirectRenderer` is not used in Mapsforge.

It was contributed by the community for a specific use case (see #1085).
It is not supported, as it is outside of Mapsforge's focus as a map library.

So how it works, you need to test and verify it. :slightly_smiling_face: 

---

Without map data, there are no generated tiles to read from outside.
Instead map server can draw its own background or 'not available' tile.

**Comment by nono303** [comment]
Thx @devemux86 for this fast and clear answer!
As `DirectRenderer` is present in Mapsforge and is not flagged unsupported (deprecated ?) We were not aware of this subtlety ;)
I understand that we have to use `DatabaseRenderer` which don't have the availability to handle 'outside of map area' for a specific action (in our case a http redirect) as it often return a non-null TileBitmap

**Comment by devemux86** [comment, marked as answer]
The meaning of unsupported is that it does not belong inside Mapsforge as it is not used by the library.
It was accepted for it to have access to other Mapsforge internal classes, but none works or maintain it.

Note that _Mapsforge is not a bitmap tile server_.

---

Mapsforge for its map views paints the outside areas with background color.
(coming from the [render theme](https://github.com/mapsforge/mapsforge/blob/master/resources/renderTheme.xsd#L304))

`DirectRenderer` (according to its authors) returns null tiles for the outside areas.
As there are no data to generate bitmap tiles for the tile server using that renderer.

**Comment by nono303** [comment]
Thx @devemux86 for these clarifications!
Adapted solution with database rendering [found](https://github.com/telemaxx/mapsforgesrv/issues/29#issuecomment-1074424310)
